feat: update rollup storage to hold pending/proven tips

just-mitch opened the pull request:
Update the rollup state to hold pending and proven tips, which themselves are structs with block/slot numbers, instead of proven/pending counts.

just-mitch committed:
update rollup storage to hold pending/proven tips

just-mitch committed:
off by 1

just-mitch committed:
update tests

just-mitch commented on l1-contracts/src/core/Rollup.sol:
```diff
@@ -462,13 +479,30 @@ contract Rollup is Leonidas, IRollup, ITestRollup {
     _validateHeader(header, _signatures, _digest, _currentTime, _txsEffectsHash, _flags);
   }
 
+  function computeTxsEffectsHash(bytes calldata _body)
```
Get my linter to stop yelling at me with respect to ordering of `external` vs `external pure`

just-mitch committed:
fix "assume proven" nonsense.

LHerskind commented on l1-contracts/src/core/Rollup.sol:
```diff
@@ -123,14 +132,13 @@ contract Rollup is Leonidas, IRollup, ITestRollup {
+    state.pendingTip = state.provenTip;
 
-    emit PrunedPending(provenBlockCount, pendingBlockCount);
+    emit PrunedPending(state.pendingTip.blockNumber);
```
Would emit both, right now you don't learn how far back you actually end up going.

LHerskind replied:
On this, the event actually say "proven" but you are using pending.

LHerskind commented on l1-contracts/src/core/Rollup.sol:
```diff
@@ -35,6 +35,16 @@ import {Leonidas} from "./sequencer_selection/Leonidas.sol";
 
+  struct ChainTip {
+    uint256 blockNumber;
+    uint256 slotNumber;
```
Why do we need the slot numbers here? 

If you have the block number, you can always go and look it up from there, so this would just add extra cost on storage.

LHerskind commented on l1-contracts/src/core/Rollup.sol:
```diff
@@ -142,8 +150,12 @@ contract Rollup is Leonidas, IRollup, ITestRollup {
-    if (blockNumber > provenBlockCount && blockNumber <= pendingBlockCount) {
-      provenBlockCount = blockNumber;
+    if (
+      blockNumber > (state.provenTip.blockNumber + 1)
```
Some of these bounds seems weird to me. If you have a pending that is proven + 1, would you not just be failing all the time.

just-mitch replied:
Agreed, they are insane. I think we should purge this function/concept altogether.

LHerskind commented on l1-contracts/src/core/Rollup.sol:
```diff
@@ -52,11 +62,9 @@ contract Rollup is Leonidas, IRollup, ITestRollup {
 
-  uint256 public pendingBlockCount;
-  uint256 public provenBlockCount;
+  State public state;
```
Would probably name this `tips` instead, since the `blocks` are also part of state, so if we want to call it state seems like we should have all of it in there.

LHerskind reviewed (commented):
Don't think the `slotNumber` in the `ChainTip` is useful here. Will add a comment on https://github.com/AztecProtocol/engineering-designs/pull/22. Can see it was also in https://github.com/AztecProtocol/engineering-designs/pull/20 so adding it there as well.

LHerskind commented:
I think this is also missing some checks that were made to the block count etc. In my mind, we would want to fix it such that we don't have the confusion of block count as well as this. And I could see at least some test still had those values.

LHerskind commented:
* **#8593** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/8593?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8583** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/8583?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @just-mitch and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

LHerskind committed:
Lh/extension chain tip (#8593)

A rebased extension of #8583. Alters the set assumes proven block as
well such that it is inclusive since it seemed weird otherwise. This
means a slight change to the environment variable as well.

LHerskind reviewed (approved):
Lets roll
